src/dtls13.c

```diff
@@ -0,0 +1,1968 @@
+            recordLength, handshake, hashOutput, 0, 0);
+        if (sendLength < 0)
+            return sendLength;
+
+        length = sendLength;
+    }
+
+    if (!FragIsInOutputBuffer(ssl, output))
+        return Dtls13SendFragFromBuffer(ssl, output, length);
+
+    ssl->buffers.outputBuffer.length += length;
+    return SendBuffered(ssl);
+}
+
+static void Dtls13FreeFragmentsBuffer(WOLFSSL *ssl)
+{
+    XFREE(ssl->dtls13FragmentsBuffer.buffer, ssl->heap,
+        DYNAMIC_TYPE_TEMP_BUFFER);
+    ssl->dtls13FragmentsBuffer.buffer = NULL;
+    ssl->dtls13SendingFragments = 0;
+    ssl->dtls13MessageLength = ssl->dtls13FragOffset = 0;
+}
+
+static WC_INLINE void Dtls13FreeRtxBufferRecord(
+    WOLFSSL *ssl, Dtls13RtxRecord *r)
+{
+    (void)ssl;
+
+    XFREE(r->data, ssl->heap, DYNAMIC_TYPE_DTLS_MSG);
+    XFREE(r, ssl->heap, DYNAMIC_TYPE_DTLS_MSG);
+}
+
+static Dtls13RtxRecord *Dtls13RtxNewRecord(WOLFSSL *ssl, byte *data,
+    word16 length, enum HandShakeType handshakeType, word32 rn[2])
+{
+    Dtls13RtxRecord *r;
+    int epochNumber;
+    byte *buf;
+
+    WOLFSSL_ENTER("Dtls13RtxNewRecord");
+
+    if (ssl->dtls13EncryptEpoch == NULL)
+        return NULL;
+
+    epochNumber = ssl->dtls13EncryptEpoch->epochNumber;
+
+    r = (Dtls13RtxRecord *)XMALLOC(
+        sizeof(*r), ssl->heap, DYNAMIC_TYPE_DTLS_MSG);
+    if (r == NULL)
+        return NULL;
+
+    buf = (byte*)XMALLOC(length, ssl->heap, DYNAMIC_TYPE_DTLS_MSG);
+    if (buf == NULL) {
+        free(r);
+        return NULL;
+    }
+
+    XMEMCPY(buf, data, length);
+
+    r->epoch = epochNumber;
+    r->data = buf;
+    r->length = length;
+    r->next = NULL;
+    r->handshakeType = handshakeType;
+    r->rn[0] = rn[0];
+    r->rn[1] = rn[1];
+    r->rnIdx = 1;
+
+    return r;
+}
+
+static void Dtls13RtxAddRecord(Dtls13RtxFSM *fsm, Dtls13RtxRecord *r)
+{
+    Dtls13RtxRecord *list;
+
+    WOLFSSL_ENTER("Dtls13RtxAddRecord");
+
+    list = fsm->rtxRecords;
+
+    /* list empty */
+    if (list == NULL) {
+        fsm->rtxRecords = r;
+        return;
+    }
+
+    while (list->next != NULL)
+        list = list->next;
+
+    list->next = r;
+}
+
+static void Dtls13RtxFlushBuffered(WOLFSSL *ssl, Dtls13RtxFSM *fsm)
+{
+    Dtls13RtxRecord *list, *r;
+
+    WOLFSSL_ENTER("Dtls13RtxFlushBuffered");
+
+    list = fsm->rtxRecords;
+
+    while (list != NULL) {
+        r = list;
+        list = r->next;
+        Dtls13FreeRtxBufferRecord(ssl, r);
+    }
+
+    fsm->rtxRecords = NULL;
+}
+
+static Dtls13RecordNumber *Dtls13NewRecordNumber(
+    WOLFSSL *ssl, word32 recordNumber[2])
+{
+    Dtls13RecordNumber *rn;
+
+    rn = (Dtls13RecordNumber *)XMALLOC(
+        sizeof(*rn), ssl->heap, DYNAMIC_TYPE_DTLS_MSG);
+    if (rn == NULL)
+        return NULL;
+
+    XMEMCPY(rn->seq, recordNumber, sizeof(rn->seq));
+    rn->next = NULL;
+
+    return rn;
+}
+
+static int Dtls13RtxAddAck(
+    WOLFSSL *ssl, Dtls13RtxFSM *fsm, word32 recordNumber[2])
```

Comment:
Be careful with array arguments. This doesn't actually mean anything other than that this is a pointer. Better to have `word32* recordNumber` with documentation. Same goes for all other array arguments.

Reply:
yeah, I wonder if record numbers deserve their own type. Epoch size is likely going to grow in the final RFC and we may want to save only the last bits of it.

Reply:
>record numbers deserve their own type

This is a good idea. Let's do that.